### PR TITLE
[feature] multiple degree matching

### PR DESCRIPTION
**Pull request by rejoycesong**
- users can now add a match degree when adding participants; they can also edit the match degree later.
- an 'Edit Participant' button and page was created; the 'DELETE' button was moved to this page. users can update the email and match degree on this page.
- updated emails to include multiple times if they were matched with multiple times
- participants are told how many times they will be assigned; given error if they select too few
- updated api
- tests for everything
- cleanup: removed the step def for programmatically generating rankings through the backend
- added stepdef for autofilling rankings

### CODE REVIEW

**Review by alined908 — APPROVED** [review]
- Multiple degree matching works as intended. 
- The edit participant page makes sense and the functionality on there is great 
- Email works as intended with multiple times, but might be good to fix the formatting to have each time be on a separate line

File: app/views/participants/show.html.erb [hunk 1]
```diff
@@ -1,49 +1,41 @@
...
+
+    <div class="match-degree">
+      <% if !@participant.match_degree == 1 %>
       <% else %>
-        <div class="form-group">
-          <td> <%= f.text_field :project_id, placeholder: "Project Id", class: "form-control"%></td>
-          <td> <%= f.text_field :email, placeholder: "Email", class: "form-control"%></td>
-          <td> <%= f.submit "Add New Participant", class:"btn btn-primary"  %></td>
-        </div>
+        <p><strong>Number of times to be assigned to you:</strong> <%= @participant.match_degree %></p>
```
Comment (kev-chien) [inline app/views/participants/show.html.erb:20]: Good cue for the participant

File: db/schema.rb [hunk 2]
```diff
@@ -23,9 +23,10 @@
     t.string   "email",          null: false
     t.datetime "last_responded"
     t.string   "secret_id"
+    t.integer  "match_degree"
   end
 
-  add_index "participants", ["project_id", "email"], name: "index_participants_on_project_id_and_email_and_last_responded", unique: true
+  add_index "participants", ["project_id", "email", "last_responded"], name: "index_participants_on_project_id_and_email_and_last_responded", unique: true
```
Comment (kev-chien) [inline db/schema.rb]: don't think this should be changed?

File: app/controllers/participants_controller.rb [hunk 3]
```diff
@@ -80,42 +81,42 @@ def handle_import
       success.empty? ? () : (flash[:success] = "Imported participants: <br/>" + success)
     end
   end
-  
+
   def handle_simple
```
Comment (kev-chien) [inline app/controllers/participants_controller.rb:85]: This is good case handling, I like the idea of defaulting to 1 match degree! (at least that's what I think this does?)

File: app/views/participants/display.html.erb [hunk 4]
```diff
@@ -25,8 +25,8 @@
           <td><%= participant.email %></td>
           <td><%= participant.last_responded ? time_ago_in_words(participant.last_responded) + ' ago' : 'No response yet'%></td>
           <td>
-            <%= link_to 'Autofill Rankings', autofill_project_participant_path(participant.project_id, participant.id), class:"btn btn-primary btn-sm"%>
-            <%= link_to 'Delete', project_participant_path(participant.project_id, participant.id), method: :delete, url: project_participants_path, class:"btn btn-danger btn-sm"%>
+            <%= link_to 'Edit Participant', project_participant_path(participant.project_id, participant.id), url: project_participants_path, class:"btn btn-info btn-sm"%>
+            <%= link_to 'Autofill Rankings', autofill_project_participant_path(participant.project_id, participant.id), class:"btn btn-primary btn-sm", id:"autofill_" + participant.id.to_s%>
```
Comment (kev-chien) [inline app/views/participants/display.html.erb:31]: I appreciate how you gave this a unique ID to make it clickable by the tests!

File: app/controllers/rankings_controller.rb [hunk 5]
```diff
@@ -60,13 +60,23 @@ def end
   # POST /rankings
   # POST /rankings.json
   def create
+    able_go_count = 0
```
Comment (kev-chien) [inline app/controllers/rankings_controller.rb:61]: Thanks for handling this check!

**Review by kev-chien — APPROVED** [review]
This is awesome and works as expected! The tests are super comprehensive, thank you so much for writing them to be not-jank, and to handle many different cases! Also thanks for fixing a lot of the previous features and tests because of this new functionality / adapting to other people's PR's.

File: app/controllers/participants_controller.rb (hunk 3 above)
Comment (rejoycesong) [reply app/controllers/participants_controller.rb:85]: Yep!

**Pull request merged**